### PR TITLE
*: report read statistics to pd 

### PR DESCRIPTION
**Pull request by nolouch**
report the read statistics.

### CODE REVIEW

File: src/coprocessor/endpoint.rs [hunk 1]
```diff
@@ -59,24 +61,46 @@ const OUTDATED_ERROR_MSG: &'static str = "request outdated.";
...
 
+pub type CopRequestStatistics = HashMap<u64, FlowStatistics>;
+
+struct CopContextFactory<R: RaftStoreRouter + 'static> {
+    raft_router: R,
+}
+
+impl<R> ContextFactory<CopContext<R>> for CopContextFactory<R>
+where
+    R: RaftStoreRouter + 'static,
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: IMO, we should not introduce RaftStoreRouter to the coprocessor, seem strange here. 

You can define a trait which sending your read stats like storage raftkv does.

File: src/raftstore/store/msg.rs [hunk 2]
```diff
@@ -75,6 +76,8 @@ pub enum Msg {
     // For snapshot stats.
     SnapshotStats,
 
+    CoprocessorStats { request_stats: CopRequestStatistics },
```
Comment (siddontang) [inline src/raftstore/store/msg.rs]: Bad style to let raftstore handle coprocessor message.

File: src/coprocessor/endpoint.rs [hunk 3]
```diff
@@ -59,24 +60,50 @@ const OUTDATED_ERROR_MSG: &'static str = "request outdated.";
...
+    pool: ThreadPool<CopContext<R>>,
+    low_priority_pool: ThreadPool<CopContext<R>>,
+    high_priority_pool: ThreadPool<CopContext<R>>,
     max_running_task_count: usize,
 }
 
+pub type CopRequestStatistics = HashMap<u64, FlowStatistics>;
+
+pub trait CopSender: Send + Clone {
+    fn send(&self, CopRequestStatistics) -> RaftStoreResult<()>;
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: Use your own Cop Result, don't use RaftStoreResult. 
IMO, you should only know RaftStore in server mod

Comment (nolouch) [reply src/coprocessor/endpoint.rs]: umm, CopResult also needs use `raftstore` mod, for implement `From` from the RaftStoreResult. I'm going to extract the PD work in another pr.

Comment (siddontang) [reply src/coprocessor/endpoint.rs]: You don't need to support From, you can use box_try or `if let Err(e) = ` + `format!` to return your own error.

**Comment by siddontang** [comment]
LGTM

But I suggest sending read statistics to the PD worker directly later. 

PTAL @BusyJay @zhangjinpeng1987

File: src/coprocessor/endpoint.rs [hunk 4]
```diff
@@ -91,9 +117,16 @@ impl CopContext {
             }
         }
     }
+
+    fn add_statistics_by_request(&mut self, id: u64, stats: &Statistics) {
+        let empty_stat = FlowStatistics::default();
+        let flow_stats = self.request_stats.entry(id).or_insert(empty_stat);
```
Comment (zhangjinpeng1987) [inline src/coprocessor/endpoint.rs]: s/empty_stat/FlowStatistics::default()

File: src/coprocessor/endpoint.rs [hunk 5]
```diff
@@ -91,9 +117,16 @@ impl CopContext {
             }
         }
     }
+
+    fn add_statistics_by_request(&mut self, id: u64, stats: &Statistics) {
```
Comment (zhangjinpeng1987) [inline src/coprocessor/endpoint.rs]: id == region_id ?

Comment (zhangjinpeng1987) [reply src/coprocessor/endpoint.rs]: s/add_statistics_by_request/add_statistics_by_region

File: src/raftstore/store/store.rs [hunk 6]
```diff
@@ -1670,13 +1686,68 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
 
-        self.register_report_region_flow_tick(event_loop);
+        self.store_stat.engine_last_total_bytes_written = 0;
+        self.store_stat.engine_last_total_keys_written = 0;
+        self.store_stat.engine_last_total_bytes_read = 0;
+        self.store_stat.engine_last_total_bytes_read = 0;
+
+    }
+
+    fn on_report_store_flow(&mut self) {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: Too complicated to understand.

File: src/server/server.rs [hunk 7]
```diff
@@ -52,6 +52,26 @@ pub struct Server<T: RaftStoreRouter + 'static, S: StoreAddrResolver + 'static>
...
+
+impl<R: RaftStoreRouter + 'static> CopReport<R> {
+    pub fn new(r: R) -> CopReport<R> {
+        CopReport { router: r.clone() }
+    }
+}
+
+impl<R: RaftStoreRouter + 'static> CopSender for CopReport<R> {
+    fn send(&self, stats: CopRequestStatistics) -> CopResult<()> {
+        box_try!(self.router.send(Msg::CoprocessorStats {
```
Comment (siddontang) [inline src/server/server.rs]: you can use try_send here

**Comment by siddontang** [comment]
LGTM

CI failed @nolouch

**Pull request merged**